package/MDAnalysis/core/topology.py

```diff
@@ -17,13 +17,74 @@
 Topology object --- :mod:`MDAnalysis.core.topology'
 ===================================================================
 
+
+"""
+
 """
+# second docstring == multiline comment
+
+TODO Notes:
+  Could make downshift tables lazily built! This would
+    a) Make these not get built when not used
+    b) Optimise moving multiple atoms between residues as only built once afterwards
+
+  Could optimise moves by only updating the two parent tables rather than rebuilding everything!
+
+
+"""
+from six.moves import zip
 import numpy as np
 
 from ..lib.mdamath import one_to_many_pointers
 from .topologyattrs import Atomindices, Resindices, Segindices
 
 
+def make_downshift_arrays(upshift):
+    """From an upwards translation table, create the opposite direction
+
+    Turns a many to one mapping (eg atoms to residues) to a one to many mapping
+    (residues to atoms)
+
+    Parameters
+    ----------
+    upshift : array_like
+        Array of integers describing which parent each item belongs to
+
+    Returns
+    -------
+    downshift : array_like (dtype object)
+        An array of arrays, each containing the indices of the children
+        of each parent
+
+    Examples
+    --------
+    
+    To find the residue to atom mappings for a given atom to residue mapping:
+
+    >>> atom2res = np.array([0, 1, 0, 2, 2, 0, 2])
+    >>> make_downshift_arrays(atom2res)
+    array([array([0, 2, 5]), array([1]), array([3, 4, 6]), None], dtype=object)
+
+    Entry 0 corresponds to residue 0 and says that this contains atoms 0, 2 & 5
+
+    Notes
+    -----
+    The final entry in the return array will be ``None`` to ensure that the
+    dtype of the array is object.  This means negative indexing should
+    not be used ever with these arrays.
+    """
+    order = np.argsort(upshift)
+
+    upshift_sorted = upshift[order]
+    borders = [None] + list(np.nonzero(np.diff(upshift_sorted))[0] + 1) + [None]
+
+    # returns an array of arrays
+    # Add None to end of array to force it to be of type Object
+    # Without this, a rectangular array gets squashed into a single array
+    return np.array([np.array(order[x:y], copy=True, dtype=np.int)
+                     for x, y in zip(borders[:-1], borders[1:])] + [None],
```

Comment:
@khuston so this is where atom order gets stored.  If this doesn't come out ordered, then we just need to change this to `np.array([np.sort(order[x:y]) for x, y etc))`